### PR TITLE
Updates to run external module versioned tests cloned from repos 

### PR DESCRIPTION
**Pull request by michaelgoin**
## Proposed Release Notes

* Runs versioned tests for external modules against tests defined in the external repository instead of tests published in npm modules.

## Details

New setup output

```

> newrelic@8.6.0 versioned:npm7
> npm run checkout-external-versioned && npm run prepare-test && NPM7=1 time ./bin/run-versioned-tests.sh


> newrelic@8.6.0 checkout-external-versioned
> node ./test/versioned-external/checkout-external-tests.js

Creating new TEMP_TESTS folder.
Executing: 'git clone --filter=blob:none --no-checkout --depth 1 --sparse https://github.com/newrelic/node-newrelic-aws-sdk.git aws-sdk'
Executing: 'git sparse-checkout set lib tests/versioned'
Executing: 'git checkout main'
Executing: 'git clone --filter=blob:none --no-checkout --depth 1 --sparse https://github.com/newrelic/node-newrelic-koa.git koa'
Executing: 'git sparse-checkout set lib tests/versioned'
Executing: 'git checkout main'
Executing: 'git clone --filter=blob:none --no-checkout --depth 1 --sparse https://github.com/newrelic/node-newrelic-superagent.git superagent'
Executing: 'git sparse-checkout set lib tests/versioned'
Executing: 'git checkout main'
```

You'll see in the test run output it is running the new test organization for aws-sdk (V2, V3) that has yet to be published with the package.

```
 ✓ amqplib (2) 13s
 ✓ bluebird (5) 33s
 ✓ cassandra-driver (2) 9s
 ✓ cls (1) 3s
 ✓ connect (6) 27s
   director
 ✓ express (16) 1m
 ✓ fastify (9) 39s
 ✓ generic-pool (2) 6s
   hapi-17
 ✓ hapi-post-18 (10) 47s
   hapi-pre-17
 ✓ ioredis (2) 8s
 ✓ koa-tests (4) 24s
 ✓ mongodb (12) 2m
 ✓ mysql (4) 36s
 ✓ mysql2 (4) 32s
 ✓ pg-post-7 (1) 18s
   pg-pre-7
 ✓ redis (2) 8s
 ✓ restify-post-7 (14) 1m
 ✓ restify-pre-7 (12) 1m
 ✓ superagent-tests (10) 37s
 ✓ undici (1) 6s
 ✓ v2 (9) 43s
 ✓ v3 (13) 2m
 ```

This will will be better in many ways and less ideal in some. Basically moves problems around so open to more discussion here or deciding we ultimately don't like this.

### Improvements

* Can stop publishing tests in external modules which bloat the install for customers. Including tests in publish is also not intuitive to those new to the agent.
* Allows test-only changes to be made in external repos due to agent changes without forcing external modules to have to be published and pulled back into the agent to get versioned tests to pass.
* Allows us to eventually test non-dependency external modules as a part of versioned runs, as desired. For example: the Apollo Server plugin or the Winston Log Enricher. (some extra work needed here).
* Allows us general flexibility in running the external tests to deal with unforseen issues (temp run against a different branch, whatever).

### Downsides

* Have to manually add configuration to add tests. Previously, any module added as a dependency would get auto-tested.
* Clones repos / requires web access every test run to get freshest tests. Alternative would be to only pull once with risk of then having outdated tests.
* Extra files cloned locally. While test files won't be duplicated once removed from external modules, the primary source (lib) files will be duplicated.

### Additional Implementation Thoughts

I tried wiring up the Apollo Server Plugin tests but ran into a funky issue. When running this way against the versioned runner... the fastify instrumentation kicked in naming the transitions differently than the tests expected. When running a single file (or directly in the Apollo repo) this doesn't happen. In the real world, I expect the fastify instrumentation will kick in. For the tests, it probably isn't loaded but unsure why kicks in for versioned runner runs? Particularly when the test (segments) is the very first ran. 🤷  Something to work out here if we want the agent to run these tests at some point (doesn't currently).

I don't think wiring up mono repo stuff (what log enrichers are becoming) if we want coverage on winston/pino. The versioned runner and versioned runner script have some limitations around path searching, so what may need to specify like a sub module path and then do some manual folder restructuring. I haven't played with this at all yet.

This highlights some funky naming with nested folders. For example: the aws-sdk tests end up with items just called v2 and v3. Maybe we should update the versioned runner to read the names in the `package.json` files so we don't have to care about folder naming/organization.

If lands, we'll want to do follow-up to remove tests from the external package's `files` list in `package.json`.

### CODE REVIEW

File: test/versioned-external/external-repos.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+'use strict'
+
+/**
+ * name: folder name to checkout the repo into.
+ * repository: repo URL to clone from.
+ * branch: branch to checkout
+ * additionalFiles: String array of files/folders to checkout in addition to lib and tests/versioned.
+ */
+const repos = [
```
Comment (bizob2828) [inline test/versioned-external/external-repos.js:14]: can you log a follow up ticket to remove the `/tests/versioned/*.js` from all the `files` arrays in package.json?

File: bin/git-commands.js [hunk 2]
```diff
@@ -86,6 +86,44 @@ async function pushTags() {
   return output
 }
 
+async function checkout(branchName) {
+  const stdout = await execAsPromise(`git checkout ${branchName}`)
```
Comment (bizob2828) [inline bin/git-commands.js:90]: so are these checkouts working because the repos are public?

File: test/versioned-external/readme.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# About versioned-external
+
+Tests from external repositories can be configured to be run by the agent as a part of the normal versioned test runs. For repos defined in `external-repos.js`, the necessary test run code will get cloned into `TEMP_TESTS` and than run as a part of the standard `npm run versioned*` test runs. These repos are cloned using sparse-checkout to attempt to reduce the amount of data pulled each time.
```
Comment (bizob2828) [inline test/versioned-external/readme.md:3]:
```suggestion
Tests from external repositories can be configured to be run by the agent as a part of the normal versioned test runs. For repos defined in `external-repos.js`, the necessary test run code will get cloned into `TEMP_TESTS` and then run as a part of the standard `npm run versioned*` test runs. These repos are cloned using sparse-checkout to attempt to reduce the amount of data pulled each time.
```

File: test/versioned-external/readme.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+Too add runs against an external repo update `external-repos.js` with the necessary details.
```
Comment (bizob2828) [inline test/versioned-external/readme.md:5]:
```suggestion
To add runs against an external repo update `external-repos.js` with the necessary details.
```

**Review by bizob2828 — APPROVED** [review]
nice work.  LGTM 🚢 🇮🇹

File: bin/git-commands.js (hunk 2 above)
Comment (michaelgoin) [reply bin/git-commands.js:90]: For CI I'd assume so as I believe the GitHub user only has immediate repo access but I didn't dig into it.

File: test/versioned-external/external-repos.js (hunk 1 above)
Comment (michaelgoin) [reply test/versioned-external/external-repos.js:14]: definitely. wanted to first make sure we were ok with the approach and maybe see it exercised a few times in earnest to make sure things work smoothly.

**Comment by michaelgoin** [comment]
@bizob2828 fixed the typos

**Review by bizob2828 — APPROVED** [review]

**Pull request merged**